Added a Kotlin version of this example app

ericmaxwell2003 opened the pull request.

ericmaxwell2003 committed:
Added Kotlin version

ericmaxwell2003 committed:
Update readme files for kotlin

bmeike commented on Kotlin/app/src/main/java/realm/io/realmsurveys/controller/QuestionViewAdapter.kt:
```diff
@@ -0,0 +1,75 @@
+
+    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
+        val view = LayoutInflater.from(parent.context).inflate(R.layout.survey_item, parent, false)
+        return ViewHolder(view)
```
Is there a reason for creating the intermediate val?

ericmaxwell2003 replied:
nope, I can combine.

bmeike commented on Kotlin/app/src/main/java/realm/io/realmsurveys/controller/QuestionViewAdapter.kt:
```diff
@@ -0,0 +1,75 @@
+        holder.noButton.setOnClickListener(clickListener)
+    }
+
+    inner class ViewHolder(val view: View) : RecyclerView.ViewHolder(view) {
```
Sometimes Kotlin is just so pretty...

bmeike commented on Kotlin/app/src/main/java/realm/io/realmsurveys/controller/SplashActivity.kt:
```diff
@@ -0,0 +1,66 @@
+
+class SplashActivity : AppCompatActivity(), AnkoLogger {
+
+    companion object {
```
I'm always a little surprised we don't put this in the Config, and get it from a build-time property...

ericmaxwell2003 replied:
I do typically, but this works better for demonstration/presentation.

bmeike commented on Kotlin/app/src/main/java/realm/io/realmsurveys/controller/SplashActivity.kt:
```diff
@@ -0,0 +1,66 @@
+
+    private fun goTo(activityClass: Class<out AppCompatActivity>) {
+        val intent = Intent(this, activityClass)
+        startActivity(intent)
```
Again... is that intermediate val necessary?

ericmaxwell2003 replied:
nope, I can combine.

bmeike commented on Kotlin/app/src/main/res/layout/activity_survey.xml:
```diff
@@ -0,0 +1,29 @@
+    app:layoutManager="LinearLayoutManager"
+    tools:context=".controller.SurveyActivity"
+    tools:listitem="@layout/survey_item"
+/>
```
CR?

ericmaxwell2003 replied:
CR?

bmeike commented on Kotlin/gradle/wrapper/gradle-wrapper.properties:
```diff
@@ -0,0 +1,6 @@
+distributionPath=wrapper/dists
+zipStoreBase=GRADLE_USER_HOME
+zipStorePath=wrapper/dists
+distributionUrl=https\://services.gradle.org/distributions/gradle-2.14.1-all.zip
```
I think the standard current version for gradle is 3.4.1

ericmaxwell2003 replied:
Thanks, good catch, will upgrade.

bmeike commented on macOS/SurveyAdmin/ViewController.swift:
```diff
@@ -23,7 +23,7 @@ struct SurveyResult {
     let noCount: Int
```
is this really a thing?  We need a swift file?

ericmaxwell2003 replied:
Sorry, messy commit.  I didn't realized I had that one in there.  It's for the MacOS version of the admin app, you can disregard this, or I can revert/commit the unchanged version to not make it part of this PR.

bmeike reviewed (approved):
Looks fine to me, if it runs.  A bunch of inconsequential comments... the only thing I *really* wish you'd do is to use a better app icon.

cmelchior commented on Kotlin/README.md:
```diff
@@ -0,0 +1,17 @@
+
+A simple survey client to receive and display questions entered in realtime and save user responses, designed to show the reactive nature of the [Realm Mobile Platform](https://realm.io/news/introducing-realm-mobile-platform/).
+
+Any number of clients may be connect to answer questions in realtime.  Answered are tallied as soon as the question is answered.
```
connected

cmelchior commented on Kotlin/README.md:
```diff
@@ -0,0 +1,17 @@
+4. Open the Android Project located at `Realm-Surveys/Android` with Android Studio
+5. Open the `SplashActivity.kt` file.  Inside there is host variable that you need to set to your Realm Object Server IP.
+6. Build the Survey app and deploy it to an Android device.
+7. When app starts you will be automatically be logged in as survey@demo.io and be able to start drawing. The Realm Object server address you enter can be local or it can be an instance running on any of our other supported Linux platforms which may also be downloaded from [Realm](https://realm.io). In either case you should ensure your firewall allows access to ports 9080 and 27800 as these are  ...
```
27800? Isn't it 7800?

ericmaxwell2003 replied:
Not sure, I took this from another one of our demo projects.  It seems to be too much information here anyway.  For a simple demo app, you're not going to be behind a firewall anyway.  You're probably running locally.

cmelchior commented on Kotlin/app/build.gradle:
```diff
@@ -0,0 +1,84 @@
+    buildTypes {
+
+        debug {
+            // This will automatically detect the IP address of the machine building RealmTasks.
```
These comments no longer apply

cmelchior commented on Kotlin/app/build.gradle:
```diff
@@ -0,0 +1,84 @@
+    testCompile 'junit:junit:4.12'
+    compile "org.jetbrains.kotlin:kotlin-stdlib-jre7:$kotlin_version"
+
+//    compile 'org.jetbrains.anko:anko-common:0.9.1'
```
Remove?

cmelchior commented on Kotlin/app/src/main/AndroidManifest.xml:
```diff
@@ -0,0 +1,40 @@
+ */
+-->
+<manifest xmlns:android="http://schemas.android.com/apk/res/android"
+    package="realm.io.realmsurveys">
```
Wrong order: It should be `io.realm.realmsurveys`

cmelchior commented on Kotlin/app/src/main/java/realm/io/realmsurveys/SurveyApplication.kt:
```diff
@@ -0,0 +1,32 @@
+ * limitations under the License.
+ */
+
+package realm.io.realmsurveys
```
All of these should be reversed to `io.realm.realmsurveys`

cmelchior commented on Kotlin/app/src/main/java/realm/io/realmsurveys/SurveyApplication.kt:
```diff
@@ -0,0 +1,32 @@
+    override fun onCreate() {
+        super.onCreate()
+        Realm.init(this)
+        RealmLog.setLevel(LogLevel.TRACE)
```
Is this really needed? It will massively flood their logs. The default is WARN, which is enough for most cases unless debugging

cmelchior commented on Java/app/src/main/java/realm/io/realmsurveys/controller/SplashActivity.java:
```diff
@@ -56,7 +56,7 @@ private void postLogin(SyncUser user) {
 
     private void setRealmDefaultConfig(SyncUser user) {
```
Why do you have both a `SplashActivity.java` and `SplashActivity.kt` ?

cmelchior commented on Kotlin/app/src/main/java/realm/io/realmsurveys/controller/SplashActivity.kt:
```diff
@@ -0,0 +1,66 @@
+        setRealmDefaultConfig(user)
+        goTo(SurveyActivity::class.java)
+    }
+
```
Whitespace

cmelchior commented on Kotlin/app/src/main/java/realm/io/realmsurveys/controller/SurveyActivity.kt:
```diff
@@ -0,0 +1,83 @@
+        setContentView(R.layout.activity_survey)
+        realm = Realm.getDefaultInstance()
+
+        val questions = realm
```
Nitpicking, but in cases like this I prefere to be able to see the structure more. I would reformat as:

```
val questions  = realm.where(Question::class.java)
                             .isEmpty("answers")
                             .or()
                             .beginGroup()
                                 .not()
                                 .contains("answers.userId", sharedPreferences.uniqueUserId())
                             .endGroup()
                             .findAllSortedAsync("timestamp") 
```

cmelchior commented on Kotlin/app/src/main/java/realm/io/realmsurveys/controller/SurveyActivity.kt:
```diff
@@ -0,0 +1,83 @@
+                question.answers?.add(answer)
+            }
+        }
+
```
Too much whitespace

cmelchior reviewed (commented):
Kotlin, yay 💃 

I would also remove the default app icons and either use our standard Realm orby (see realm-java/examples) or the rather nice playstore icon you already have.

ericmaxwell2003 committed:
Undo IP change in admin client

ericmaxwell2003 committed:
Remove IP change in Java version

ericmaxwell2003 committed:
Remove Schema version change in Java version

ericmaxwell2003 committed:
Fix verbiage

ericmaxwell2003 committed:
Remove superflous instructions

ericmaxwell2003 committed:
Clean up build.gradle comments and spacing

ericmaxwell2003 committed:
Change package naming

ericmaxwell2003 committed:
Merge branch 'em/addKotlinVersion' of https://github.com/realm-demos/realm-surveys into em/addKotlinVersion

ericmaxwell2003 committed:
Replace the launch icon as requested in PR

ericmaxwell2003 committed:
Remove accidental change to SurveyAdmin/ViewController

ericmaxwell2003 commented:
I made the updates requested mostly.
- @bmeike - Once I chained a few of those calls into one line instead of multiple variables, I didn't like the way the code read, so I put them back.  For example 
```java
      val syncConfiguration = SyncConfiguration.Builder(user, REALM_URL).build()
      Realm.setDefaultConfiguration(syncConfiguration)
```
I added all the carriage returns and formatted the layouts to make the spacing more standard.

@cmelchior , good catch on the package naming, I totally flipped those.  I actually did that in the original java version as well, so I will go back and update that after this.

I updated the launcher icon.  Looks much nicer now, good suggestion.  

All good?
